File: drop-ins/go/feedback_tool.go
// PatchworkMCP — Drop-in feedback tool for Go MCP servers.
//
// Copy this file into your project. Works with:
//   - github.com/mark3labs/mcp-go  → RegisterFeedbackTool(server, "my-server")
//   - Manual registration          → NewFeedbackTool(), NewFeedbackHandler()
//
// No extra dependencies beyond mcp-go and the standard library.
//
// Configuration via environment:
//   FEEDBACK_SIDECAR_URL  - default: http://localhost:8099
//   FEEDBACK_API_KEY      - optional shared secret

package feedback

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"math"
	"net"
	"net/http"
	"os"
	"time"

	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
)

var (
	sidecarURL = getEnv("FEEDBACK_SIDECAR_URL", "http://localhost:8099")
	apiKey     = os.Getenv("FEEDBACK_API_KEY")
)

// ── HTTP Client Config ─────────────────────────────────────────────────────

const (
	maxRetries     = 2
	initialBackoff = 500 * time.Millisecond // doubles each retry
	userAgent      = "PatchworkMCP-Go/1.0"
)

// Module-level client with connection pooling and sensible timeouts.
var httpClient = &http.Client{
	Timeout: 5 * time.Second,
	Transport: &http.Transport{
		DialContext: (&net.Dialer{
			Timeout: 2 * time.Second,
		}).DialContext,
		MaxIdleConns:        10,
		MaxIdleConnsPerHost: 5,
		IdleConnTimeout:     90 * time.Second,
	},
}

// Prefix makes these log lines greppable in any log aggregator.
const logPrefix = "PATCHWORKMCP_UNSENT_FEEDBACK"

func isRetryableStatus(code int) bool {
	return code == 429 || code == 500 || code == 502 || code == 503 || code == 504
}

// logUnsentPayload writes the full payload to stderr at warning level so the
// hosting environment captures it. The structured JSON is greppable via
// logPrefix and can be replayed from whatever log aggregation the containing
// server uses (Heroku logs, CloudWatch, Docker stdout, etc.).
func logUnsentPayload(body []byte, reason string) {
	fmt.Fprintf(os.Stderr, "%s reason=%s payload=%s\n", logPrefix, reason, string(body))
}

func getEnv(key, fallback string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return fallback
}

// ── Tool Schema ─────────────────────────────────────────────────────────────

const ToolName = "feedback"

const ToolDescription = "Report when you cannot find what you need or when available tools don't " +
	"fully address the task. This feedback directly improves this server. " +
	"Call this tool whenever: " +
	"(1) you looked for a tool or resource that doesn't exist, " +
	"(2) a tool returned incomplete or unhelpful results, " +
	"(3) you had to work around a limitation or approximate an answer, " +
	"(4) a new tool or parameter would have made the task easier. " +
	"If you could not fully satisfy the user's request with the available " +
	"tools, call this BEFORE giving your final response."

// NewFeedbackTool returns the MCP tool definition for registration.
func NewFeedbackTool() mcp.Tool {
	return mcp.NewTool(ToolName,
		mcp.WithDescription(ToolDescription),
		mcp.WithString("what_i_needed",
			mcp.Required(),
			mcp.Description("What capability, data, or tool were you looking for?"),
		),
		mcp.WithString("what_i_tried",
			mcp.Required(),
			mcp.Description("What tools or approaches did you try? Include tool names and brief results."),
		),
		mcp.WithString("gap_type",
			mcp.Required(),
			mcp.Description("The category of gap: missing_tool, incomplete_results, missing_parameter, wrong_format, other"),
		),
		mcp.WithString("suggestion",
			mcp.Description("Your idea for what would have helped — inputs, outputs, behavior."),
		),
		mcp.WithString("user_goal",
			mcp.Description("The user's original request that led to discovering this gap."),
		),
		mcp.WithString("resolution",
			mcp.Description("What happened after the gap: blocked, worked_around, partial"),
		),
		mcp.WithString("agent_model",
			mcp.Description("Your model identifier, if known."),
		),
		mcp.WithString("session_id",
			mcp.Description("An identifier for the current conversation or session."),
		),
		mcp.WithString("client_type",
			mcp.Description("The MCP client in use, if known (e.g. 'claude-desktop', 'cursor', 'claude-code')."),
		),
		// Note: tools_available is sent as a comma-separated string in the Go
		// drop-in since mcp-go doesn't have WithArray. The sidecar accepts both
		// array and string formats.
		mcp.WithString("tools_available",
			mcp.Description("Comma-separated list of tool names you considered or tried."),
		),
	)
}

// ── Feedback Submission ─────────────────────────────────────────────────────

type feedbackPayload struct {
	ServerName   string   `json:"server_name"`
	WhatINeeded  string   `json:"what_i_needed"`
	WhatITried   string   `json:"what_i_tried"`
	GapType      string   `json:"gap_type"`
	Suggestion   string   `json:"suggestion"`
	UserGoal     string   `json:"user_goal"`
	Resolution   string   `json:"resolution"`
	AgentModel   string   `json:"agent_model"`
	SessionID    string   `json:"session_id"`
	ClientType   string   `json:"client_type"`
	ToolsAvail   []string `json:"tools_available"`
}

func getString(args map[string]any, key string) string {
	if v, ok := args[key]; ok {
		if s, ok := v.(string); ok {
			return s
		}
	}
	return ""
}

// Options configures the feedback tool's sidecar connection.
// Pass to RegisterFeedbackTool or SendFeedback to override env vars.
type Options struct {
	// SidecarURL overrides FEEDBACK_SIDECAR_URL.
	SidecarURL string
	// APIKey overrides FEEDBACK_API_KEY.
	APIKey string
}

func (o *Options) url() string {
	if o != nil && o.SidecarURL != "" {
		return o.SidecarURL
	}
	return sidecarURL
}

func (o *Options) key() string {
	if o != nil && o.APIKey != "" {
		return o.APIKey
	}
	return apiKey
}

// SendFeedback posts feedback to the sidecar with retry logic.
//
// Retries up to maxRetries times on transient failures (connection errors,
// 5xx, 429) with exponential backoff. Uses a module-level http.Client for
// connection pooling. Best-effort — returns a message regardless of outcome.
// Pass nil for opts to use environment variable defaults.
func SendFeedback(ctx context.Context, args map[string]any, serverName string, opts *Options) string {
	// Parse tools_available — accept comma-separated string or []any
	var tools []string
	switch v := args["tools_available"].(type) {
	case string:
		if v != "" {
			for _, t := range bytes.Split([]byte(v), []byte(",")) {
				tools = append(tools, string(bytes.TrimSpace(t)))
			}
		}
	case []any:
		for _, t := range v {
			if s, ok := t.(string); ok {
				tools = append(tools, s)
			}
		}
	}

	payload := feedbackPayload{
		ServerName:  serverName,
		WhatINeeded: getString(args, "what_i_needed"),
		WhatITried:  getString(args, "what_i_tried"),
		GapType:     getString(args, "gap_type"),
		Suggestion:  getString(args, "suggestion"),
		UserGoal:    getString(args, "user_goal"),
		Resolution:  getString(args, "resolution"),
		AgentModel:  getString(args, "agent_model"),
		SessionID:   getString(args, "session_id"),
		ClientType:  getString(args, "client_type"),
		ToolsAvail:  tools,
	}
	if payload.GapType == "" {
		payload.GapType = "other"
	}

	body, err := json.Marshal(payload)
	if err != nil {
		return "Feedback noted (encoding error)."
	}

	endpoint := opts.url() + "/api/feedback"
	authKey := opts.key()
	var lastErr error

	for attempt := 0; attempt <= maxRetries; attempt++ {
		req, err := http.NewRequestWithContext(ctx, "POST", endpoint, bytes.NewReader(body))
		if err != nil {
			return "Feedback noted (sidecar unavailable, but your input is appreciated)."
		}
		req.Header.Set("Content-Type", "application/json")
		req.Header.Set("User-Agent", userAgent)
		if authKey != "" {
			req.Header.Set("Authorization", "Bearer "+authKey)
		}

		resp, err := httpClient.Do(req)
		if err != nil {
			lastErr = err
			if attempt < maxRetries {
				backoff := time.Duration(float64(initialBackoff) * math.Pow(2, float64(attempt)))
				select {
				case <-ctx.Done():
					logUnsentPayload(body, fmt.Sprintf("unreachable:%v", err))
					return "Feedback could not be delivered and was logged. (Server unreachable)"
				case <-time.After(backoff):
				}
				continue
			}
			break
		}
		// Drain body so the connection can be reused.
		io.Copy(io.Discard, resp.Body)
		resp.Body.Close()

		if resp.StatusCode == 201 {
			return "Thank you. Your feedback has been recorded and will be used to improve this server's capabilities."
		}
		if isRetryableStatus(resp.StatusCode) && attempt < maxRetries {
			backoff := time.Duration(float64(initialBackoff) * math.Pow(2, float64(attempt)))
			select {
			case <-ctx.Done():
				logUnsentPayload(body, fmt.Sprintf("status_%d", resp.StatusCode))
				return fmt.Sprintf("Feedback could not be delivered and was logged. (Server returned %d)", resp.StatusCode)
			case <-time.After(backoff):
			}
			continue
		}
		logUnsentPayload(body, fmt.Sprintf("status_%d", resp.StatusCode))
		return fmt.Sprintf("Feedback could not be delivered and was logged. (Server returned %d)", resp.StatusCode)
	}

	logUnsentPayload(body, fmt.Sprintf("unreachable:%v", lastErr))
	return "Feedback could not be delivered and was logged. (Server unreachable)"
}

// ── Handler & Registration ──────────────────────────────────────────────────

// NewFeedbackHandler returns a tool handler function bound to a server name.
// Pass nil for opts to use environment variable defaults.
func NewFeedbackHandler(serverName string, opts *Options) server.ToolHandlerFunc {
	return func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		args := req.GetArguments()
		msg := SendFeedback(ctx, args, serverName, opts)
		return mcp.NewToolResultText(msg), nil
	}
}

// RegisterFeedbackTool is a one-liner to add the feedback tool to an MCP server.
// Pass nil for opts to use environment variable defaults.
//
//	s := server.NewMCPServer("my-server", "1.0.0")
//	feedback.RegisterFeedbackTool(s, "my-server", nil)
//
//	// Or point at a specific sidecar:
//	feedback.RegisterFeedbackTool(s, "my-server", &feedback.Options{
//	    SidecarURL: "https://feedback.prod.example.com",
//	})
func RegisterFeedbackTool(s *server.MCPServer, serverName string, opts *Options) {
	s.AddTool(NewFeedbackTool(), NewFeedbackHandler(serverName, opts))
}

File: drop-ins/rust/feedback_tool.rs
//! PatchworkMCP — Drop-in feedback tool for Rust MCP servers.
//!
//! Copy this file into your project and call `register_feedback_tool()` with
//! your MCP server router, or use the constants and `send_feedback()` directly.
//!
//! Dependencies (add to Cargo.toml):
//!   reqwest = { version = "0.12", features = ["json"] }
//!   serde = { version = "1", features = ["derive"] }
//!   serde_json = "1"
//!   tokio = { version = "1", features = ["full"] }
//!
//! Configuration via environment:
//!   FEEDBACK_SIDECAR_URL  - default: http://localhost:8099
//!   FEEDBACK_API_KEY      - optional shared secret
//!
//! Note: The Rust MCP ecosystem is still maturing. This file provides the
//! feedback payload, HTTP submission, and schema constants. Wire the tool
//! into your MCP framework's registration system as needed.

use serde::{Deserialize, Serialize};
use std::env;
use std::time::Duration;

// ── Constants ───────────────────────────────────────────────────────────────

pub const TOOL_NAME: &str = "feedback";

pub const TOOL_DESCRIPTION: &str = concat!(
    "Report when you cannot find what you need or when available tools don't ",
    "fully address the task. This feedback directly improves this server. ",
    "Call this tool whenever: ",
    "(1) you looked for a tool or resource that doesn't exist, ",
    "(2) a tool returned incomplete or unhelpful results, ",
    "(3) you had to work around a limitation or approximate an answer, ",
    "(4) a new tool or parameter would have made the task easier. ",
    "If you could not fully satisfy the user's request with the available ",
    "tools, call this BEFORE giving your final response.",
);

// ── Types ───────────────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FeedbackPayload {
    pub server_name: String,
    pub what_i_needed: String,
    pub what_i_tried: String,
    pub gap_type: String,
    #[serde(default)]
    pub suggestion: String,
    #[serde(default)]
    pub user_goal: String,
    #[serde(default)]
    pub resolution: String,
    #[serde(default)]
    pub agent_model: String,
    #[serde(default)]
    pub session_id: String,
    #[serde(default)]
    pub tools_available: Vec<String>,
}

#[derive(Debug, Deserialize)]
struct SidecarResponse {
    #[allow(dead_code)]
    id: String,
    #[allow(dead_code)]
    status: String,
}

// ── Config ──────────────────────────────────────────────────────────────────

fn sidecar_url() -> String {
    env::var("FEEDBACK_SIDECAR_URL").unwrap_or_else(|_| "http://localhost:8099".to_string())
}

fn api_key() -> Option<String> {
    env::var("FEEDBACK_API_KEY").ok().filter(|k| !k.is_empty())
}

// ── Submission ──────────────────────────────────────────────────────────────

/// Send feedback to the PatchworkMCP sidecar. Best-effort — returns a
/// user-facing message regardless of success or failure.
pub async fn send_feedback(payload: &FeedbackPayload) -> String {
    let url = format!("{}/api/feedback", sidecar_url());

    let client = match reqwest::Client::builder()
        .timeout(Duration::from_secs(5))
        .build()
    {
        Ok(c) => c,
        Err(_) => {
            return "Feedback noted (HTTP client error).".to_string();
        }
    };

    let mut req = client.post(&url).json(payload);
    if let Some(key) = api_key() {
        req = req.header("Authorization", format!("Bearer {key}"));
    }

    match req.send().await {
        Ok(resp) if resp.status().as_u16() == 201 => {
            "Thank you. Your feedback has been recorded and will be \
             used to improve this server's capabilities."
                .to_string()
        }
        Ok(resp) => {
            eprintln!(
                "PatchworkMCP sidecar returned {}",
                resp.status().as_u16()
            );
            "Feedback noted (delivery issue, but recorded locally).".to_string()
        }
        Err(e) => {
            eprintln!("PatchworkMCP: could not reach sidecar: {e}");
            "Feedback noted (sidecar unavailable, but your input is appreciated).".to_string()
        }
    }
}

/// Build a FeedbackPayload from a JSON value (as received from MCP call_tool).
/// Missing fields get sensible defaults.
pub fn payload_from_args(args: &serde_json::Value, server_name: &str) -> FeedbackPayload {
    let s = |key: &str| -> String {
        args.get(key)
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string()
    };

    let tools: Vec<String> = args
        .get("tools_available")
        .and_then(|v| v.as_array())
        .map(|arr| {
            arr.iter()
                .filter_map(|v| v.as_str().map(String::from))
                .collect()
        })
        .unwrap_or_default();

    FeedbackPayload {
        server_name: server_name.to_string(),
        what_i_needed: s("what_i_needed"),
        what_i_tried: s("what_i_tried"),
        gap_type: {
            let g = s("gap_type");
            if g.is_empty() {
                "other".to_string()
            } else {
                g
            }
        },
        suggestion: s("suggestion"),
        user_goal: s("user_goal"),
        resolution: s("resolution"),
        agent_model: s("agent_model"),
        session_id: s("session_id"),
        tools_available: tools,
    }
}

// ── JSON Schema (for manual tool registration) ──────────────────────────────

/// Returns the tool input schema as a serde_json::Value. Use this when
/// registering the tool manually with your MCP framework.
pub fn tool_input_schema() -> serde_json::Value {
    serde_json::json!({
        "type": "object",
        "properties": {
            "what_i_needed": {
                "type": "string",
                "description": "What capability, data, or tool were you looking for?"
            },
            "what_i_tried": {
                "type": "string",
                "description": "What tools or approaches did you try? Include tool names and brief results."
            },
            "gap_type": {
                "type": "string",
                "enum": ["missing_tool", "incomplete_results", "missing_parameter", "wrong_format", "other"],
                "description": "The category of gap encountered."
            },
            "suggestion": {
                "type": "string",
                "description": "Your idea for what would have helped."
            },
            "user_goal": {
                "type": "string",
                "description": "The user's original request or goal."
            },
            "resolution": {
                "type": "string",
                "enum": ["blocked", "worked_around", "partial"],
                "description": "What happened after hitting the gap."
            },
            "tools_available": {
                "type": "array",
                "items": { "type": "string" },
                "description": "Tool names you considered or tried."
            },
            "agent_model": {
                "type": "string",
                "description": "Your model identifier, if known."
            },
            "session_id": {
                "type": "string",
                "description": "Conversation or session identifier."
            }
        },
        "required": ["what_i_needed", "what_i_tried", "gap_type"]
    })
}
